### PR TITLE
Exllama q4 kernel

### PR DESCRIPTION
**Pull request by qwopqwop200**
1. Add pack function (Follow cuda pack function.)
2. Since the input of exllama must be float16, it has been changed to automatically convert to float16.
3. exllama does not support backpropagation, so I changed to automatically disable when training.

### CODE REVIEW

File: auto_gptq/nn_modules/qlinear/qlinear_exllama.py [hunk 1]
```diff
@@ -93,10 +97,107 @@ def post_init(self):
...
+                i += 10
+                col += 1
+            else:
+                raise NotImplementedError("Only 2,3,4,8 bits are supported.")
+
+        qzeros = qzeros.astype(np.int32)
+        self.qzeros = torch.from_numpy(qzeros)
 
     def forward(self, x):
-        out = ext_q4_matmul(x, self.q4, self.width)
```
Comment (fxmarty) [inline auto_gptq/nn_modules/qlinear/qlinear_exllama.py:99]: The input here should already be in fp16 no? Do you know if the other implems (cuda, triton) support fp32?

File: auto_gptq/nn_modules/qlinear/qlinear_exllama.py [hunk 2]
```diff
@@ -93,10 +97,107 @@ def post_init(self):
...
+                qzeros[:, col] |= zeros[:, i] << 31
+                col += 1
+                qzeros[:, col] |= (zeros[:, i] >> 1) & 0x3
+                i += 1
+                for j in range(i, i + 10):
+                    qzeros[:, col] |= zeros[:, j] << (3 * (j - i) + 2)
+                i += 10
+                col += 1
+            else:
+                raise NotImplementedError("Only 2,3,4,8 bits are supported.")
```
Comment (fxmarty) [inline auto_gptq/nn_modules/qlinear/qlinear_exllama.py]: Where is this `pack` function used? For exllama, the kernel can handle only 4-bit quantization, not sure if it is an issue here.

**Review by fxmarty — COMMENTED** [review]
Awesome, I left two comments!

File: auto_gptq/nn_modules/qlinear/qlinear_exllama.py (hunk 2 above)
Comment (qwopqwop200) [reply auto_gptq/nn_modules/qlinear/qlinear_exllama.py]: The pack function is used in the process of converting quantized float values ​​(fake quantization) to int values ​​(real quantization).
And since exllama only works with 4-bit quantization, the pack function other than 4-bit has been removed.

File: auto_gptq/nn_modules/qlinear/qlinear_exllama.py (hunk 1 above)
Comment (qwopqwop200) [reply auto_gptq/nn_modules/qlinear/qlinear_exllama.py:99]: For cuda, it is automatically converted to float. https://github.com/PanQiWei/AutoGPTQ/blob/main/auto_gptq/nn_modules/qlinear/qlinear_cuda.py#L203
In the case of triton, it changes automatically based on input.
https://github.com/PanQiWei/AutoGPTQ/blob/main/auto_gptq/nn_modules/triton_utils/kernels.py#L321

**Comment by qwopqwop200** [comment]
Added fused attention for llama 70b.
exllama only supports flash attention.

File: auto_gptq/nn_modules/fused_llama_attn.py [hunk 3]
```diff
@@ -39,24 +55,32 @@ def _shape(self, tensor, seq_len, bsz):
...
+            key_states = self.k_proj(hidden_states)
+            value_states = self.v_proj(hidden_states)
         query_states = query_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
-        key_states = key_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
-        value_states = value_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
-
+        key_states = key_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
+        value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
+        
         kv_seq_len = key_states.shape[-2]
```
Comment (fxmarty) [inline auto_gptq/nn_modules/fused_llama_attn.py]: This seem unrelated to this PR no?

File: auto_gptq/nn_modules/fused_llama_attn.py [hunk 4]
```diff
@@ -79,6 +103,10 @@ def forward(
 
         past_key_value = (key_states, value_states) if use_cache else None
 
+        # repeat k/v heads if n_kv_heads < n_heads
+        key_states = repeat_kv(key_states, self.num_key_value_groups)
+        value_states = repeat_kv(value_states, self.num_key_value_groups)
+
```
Comment (fxmarty) [inline auto_gptq/nn_modules/fused_llama_attn.py]: same

File: auto_gptq/nn_modules/fused_llama_attn.py [hunk 5]
```diff
@@ -150,41 +177,63 @@ def inject_to_model(
...
-
-            if QuantLinear.QUANT_TYPE == "exllama":
-                if desc_act:
-                    # TODO: support it. The issue lies maybe in the line:
-                    # int groups = qzeros.size(0);
-                    # in exllama_ext.cpp
-                    raise ValueError("Exllama kernel does not support query/key/value fusion with act-order. Please either use inject_fused_attention=False or disable_exllama=True.")
-                else:
+            if QuantLinear.QUANT_TYPE == "exllama" and desc_act:
+                qkv_layers = [q_proj,k_proj,v_proj]
```
Comment (fxmarty) [inline auto_gptq/nn_modules/fused_llama_attn.py]: Is this really working? Did you verify correctness? I was thinking it is more of an issue in the CUDA kernel.

**Comment by fxmarty** [comment]
@qwopqwop200 Supporting grouped query attention is unrelated, could we rather do that in an other PR?

**Comment by qwopqwop200** [comment]
ok

**Review by fxmarty — APPROVED** [review]
Thanks a lot @qwopqwop200!
